planner/core/optimizer.go

```diff
@@ -386,6 +387,106 @@ func postOptimize(sctx sessionctx.Context, plan PhysicalPlan) PhysicalPlan {
+	emptyColumns := make([]*expression.Column, 0)
+	if tableReader, ok := plan.(*PhysicalTableReader); ok {
+		if _, isExchangeSender := tableReader.tablePlan.(*PhysicalExchangeSender); isExchangeSender {
+			prunePhysicalColumnsInternal(sctx, tableReader.tablePlan, emptyColumns, false)
+		}
+	} else {
+		for _, child := range plan.Children() {
+			prunePhysicalColumns(sctx, child)
+		}
+	}
+}
+
+func (p *PhysicalHashJoin) extractUsedCols(parentUsedCols []*expression.Column) (leftCols []*expression.Column, rightCols []*expression.Column) {
+	for _, eqCond := range p.EqualConditions {
+		parentUsedCols = append(parentUsedCols, expression.ExtractColumns(eqCond)...)
+	}
+	for _, neCond := range p.NAEqualConditions {
+		parentUsedCols = append(parentUsedCols, expression.ExtractColumns(neCond)...)
+	}
+	for _, leftCond := range p.LeftConditions {
+		parentUsedCols = append(parentUsedCols, expression.ExtractColumns(leftCond)...)
+	}
+	for _, rightCond := range p.RightConditions {
+		parentUsedCols = append(parentUsedCols, expression.ExtractColumns(rightCond)...)
+	}
+	for _, otherCond := range p.OtherConditions {
+		parentUsedCols = append(parentUsedCols, expression.ExtractColumns(otherCond)...)
+	}
+	lChild := p.children[0]
+	rChild := p.children[1]
+	for _, col := range parentUsedCols {
+		if lChild.Schema().Contains(col) {
+			leftCols = append(leftCols, col)
+		} else if rChild.Schema().Contains(col) {
+			rightCols = append(rightCols, col)
+		}
+	}
+	return leftCols, rightCols
+}
+
+func prunePhysicalColumnsInternal(sctx sessionctx.Context, plan PhysicalPlan, parentUsedCols []*expression.Column, meaningFulParentUsedCols bool) {
+	switch x := plan.(type) {
+	case *PhysicalHashJoin:
+		child0 := x.children[0]
+		child1 := x.children[1]
+		schemaColumns := x.Schema().Clone().Columns
+		leftCols, rightCols := x.extractUsedCols(schemaColumns)
+		prunePhysicalColumnsInternal(sctx, child0, leftCols, true)
+		prunePhysicalColumnsInternal(sctx, child1, rightCols, true)
+	case *PhysicalExchangeReceiver:
+		// Currently, prune columns only when parent node is HashJoin
+		if meaningFulParentUsedCols {
+			switch y := x.children[0].(type) {
+			case *PhysicalExchangeSender:
+				used := expression.GetUsedList(parentUsedCols, y.Schema())
+
+				var exprCols []*expression.Column
+				for _, mppCol := range y.HashCols {
+					exprCols = append(exprCols, mppCol.Col)
+				}
+				exprUsed := expression.GetUsedList(exprCols, y.Schema())
+
+				needPrune := false
+				usedExprs := make([]expression.Expression, len(y.Schema().Columns))
+				prunedSchema := y.Schema().Clone()
+				for i := len(used) - 1; i >= 0; i-- {
+					usedExprs[i] = x.Schema().Columns[i]
+					if !used[i] && !exprUsed[i] {
+						needPrune = true
+						usedExprs = append(usedExprs[:i], usedExprs[i+1:]...)
+						prunedSchema.Columns = append(prunedSchema.Columns[:i], prunedSchema.Columns[i+1:]...)
+					}
+				}
+
+				if needPrune {
+					ch := y.children[0]
```

Comment:
Do we need to recursively prune columns for `ch`?

Reply:
Good catch, fix it.